patches/minecraft/net/minecraft/network/NetHandlerPlayServer.java.patch

```diff
@@ -1,6 +1,37 @@
 --- ../src-base/minecraft/net/minecraft/network/NetHandlerPlayServer.java
 +++ ../src-work/minecraft/net/minecraft/network/NetHandlerPlayServer.java
-@@ -611,7 +611,10 @@
+@@ -90,7 +90,9 @@
+ import net.minecraft.tileentity.TileEntityCommandBlock;
+ import net.minecraft.tileentity.TileEntitySign;
+ import net.minecraft.tileentity.TileEntityStructure;
++import net.minecraft.util.ActionResult;
+ import net.minecraft.util.ChatAllowedCharacters;
++import net.minecraft.util.EnumActionResult;
+ import net.minecraft.util.EnumFacing;
+ import net.minecraft.util.EnumHand;
+ import net.minecraft.util.ITickable;
+@@ -571,9 +573,17 @@
+ 
+                 if (!this.field_147369_b.func_175149_v())
+                 {
+-                    ItemStack itemstack1 = this.field_147369_b.func_184586_b(EnumHand.OFF_HAND);
+-                    this.field_147369_b.func_184611_a(EnumHand.OFF_HAND, this.field_147369_b.func_184586_b(EnumHand.MAIN_HAND));
+-                    this.field_147369_b.func_184611_a(EnumHand.MAIN_HAND, itemstack1);
++                	ItemStack itemstack1 = this.field_147369_b.func_184586_b(EnumHand.OFF_HAND);
++                    ItemStack itemstack2 = this.field_147369_b.func_184586_b(EnumHand.MAIN_HAND);
++                    
++                    ActionResult<ItemStack> res1 = itemstack1 != null ? itemstack1.func_77973_b().onSwapHand(this.field_147369_b, EnumHand.OFF_HAND, itemstack1, EnumHand.MAIN_HAND, itemstack2) : new ActionResult(EnumActionResult.PASS, null);
++                    ActionResult<ItemStack> res2 = itemstack2 != null ? itemstack2.func_77973_b().onSwapHand(this.field_147369_b, EnumHand.MAIN_HAND, itemstack2, EnumHand.OFF_HAND, itemstack1) : new ActionResult(EnumActionResult.PASS, null);
++                    
++                    if(res1.func_188397_a() == EnumActionResult.PASS && res2.func_188397_a() == EnumActionResult.PASS)
++                    {
++                    	this.field_147369_b.func_184611_a(EnumHand.OFF_HAND, res2.func_188398_b());
++                    	this.field_147369_b.func_184611_a(EnumHand.MAIN_HAND, res1.func_188398_b());
++                    }
```

Comment:
This whole ordeal can be simplified to a 2 line patch. Externalize all the checks into ForgeHooks and don't change the indentation on vanilla code.